app/src/main/java/ro/code4/monitorizarevot/ui/login/LoginActivity.kt

```diff
@@ -29,13 +40,49 @@ class LoginActivity : BaseActivity<LoginViewModel>() {
             )//TODO replace with phone uiid  Settings.Secure.getString(getContentResolver(), Settings.Secure.ANDROID_ID);
             viewModel.login(user)
         }
-        appVersion.text = getString(R.string.app_version, BuildConfig.VERSION_NAME)
+    }
+
+    private fun observablesSetup() {
+        loginUserObservable()
+        onboardingObservable()
+    }
+
+    private fun loginUserObservable() {
         viewModel.loggedIn().observe(this, Observer {
-            startActivityWithoutTrace(BranchActivity::class.java)
+            it.handle(
+                onSuccess = {
+                    progressDialog.dismiss()
+                    startActivityWithoutTrace(BranchActivity::class.java)
+                },
+                onFailure = {
+                    // TODO: Handle errors to show personalized messages for each one
+                    progressDialog.dismiss()
+                    Snackbar.make(loginButton, "Something went wrong!", Snackbar.LENGTH_SHORT)
+                        .show()
+                },
+                onLoading = {
+                    progressDialog.show(supportFragmentManager, ProgressDialogFragment.tag)
+                }
+            )
         })
+    }
+
+    private fun onboardingObservable() {
         viewModel.onboarding().observe(this, Observer {
-            startActivityWithoutTrace(OnboardingActivity::class.java)
+            it.handle(
+                onSuccess = {
+                    progressDialog.dismiss()
+                    startActivityWithoutTrace(OnboardingActivity::class.java)
+                },
+                onFailure = {
+                    progressDialog.dismiss()
+                    Snackbar.make(loginButton, "Something went wrong!", Snackbar.LENGTH_SHORT)
+                        .show()
+                },
+                onLoading = {
+                    progressDialog.show(supportFragmentManager, ProgressDialogFragment.tag)
+                }
+            )
```

Comment:
The only difference between these is the next activity that should be started if the user logged in successfully. You could implement a method that has the activity as a parameter.